pkg/tfshim/sdk-v2/proposed_new.go

```diff
@@ -52,39 +47,189 @@ func htype2ctype(t hcty.Type) (cty.Type, error) {
-		return cty.NilVal, nil
-	}
-	typ, err := htype2ctype(v.Type())
-	if err != nil {
-		return cty.False, err
-	}
-	valueJSON, err := hctyjson.Marshal(v, v.Type())
-	if err != nil {
-		return cty.False, err
+func hcty2cty(val hcty.Value) cty.Value {
+	if val.IsKnown() && val.Equals(hcty.NilVal).True() {
+		return cty.NilVal
 	}
-	return ctyjson.Unmarshal(valueJSON, typ)
+	ty, err := htype2ctype(val.Type())
+	contract.AssertNoErrorf(err, "unexpected error in htype2ctype")
+	return hcty2ctyWithType(ty, val)
 }
 
-func cty2hcty(v cty.Value) (hcty.Value, error) {
-	if v.Equals(cty.NilVal).True() {
-		return hcty.NilVal, nil
+func hcty2ctyWithType(ty cty.Type, val hcty.Value) cty.Value {
+	// Pattern-matching below is based on how hcty.Transform is written.
+	switch {
+	case val.IsNull():
+		return cty.NullVal(ty)
+	case !val.IsKnown():
+		return cty.UnknownVal(ty)
+	case ty.Equals(cty.EmptyTuple):
+		return cty.EmptyTupleVal
+	case ty.Equals(cty.String):
+		return cty.StringVal(val.AsString())
+	case ty.Equals(cty.Number):
+		return cty.NumberVal(val.AsBigFloat())
+	case ty.Equals(cty.Bool):
+		if val.True() {
+			return cty.True
+		}
+		return cty.False
+	case ty.IsListType(), ty.IsSetType(), ty.IsTupleType():
+		l := val.LengthInt()
+		elems := make([]cty.Value, 0, l)
+		idx := 0
+		for it := val.ElementIterator(); it.Next(); {
+			_, ev := it.Element()
+			var eT cty.Type
+			if ty.IsTupleType() {
+				eT = ty.TupleElementType(idx)
+			} else {
+				eT = ty.ElementType()
+			}
+			newEv := hcty2ctyWithType(eT, ev)
+			elems = append(elems, newEv)
+			idx++
+		}
+		switch {
+		case ty.IsListType():
+			if len(elems) == 0 {
+				return cty.ListValEmpty(ty.ElementType())
+			}
+			return cty.ListVal(elems)
+		case ty.IsSetType():
+			if len(elems) == 0 {
+				return cty.SetValEmpty(ty.ElementType())
+			}
+			return cty.SetVal(elems)
+		case ty.IsTupleType():
+			return cty.TupleVal(elems)
+		default:
+			contract.Assertf(false, "match failure on cty.Type: unknown sequence type")
+		}
+	case ty.IsMapType():
+		eT := ty.ElementType()
+		elems := make(map[string]cty.Value)
+		for it := val.ElementIterator(); it.Next(); {
+			kv, ev := it.Element()
+			newEv := hcty2ctyWithType(eT, ev)
+			elems[kv.AsString()] = newEv
+		}
+		if len(elems) == 0 {
+			return cty.MapValEmpty(eT)
+		}
+		return cty.MapVal(elems)
+	case ty.IsObjectType():
+		atys := ty.AttributeTypes()
+		newAVs := make(map[string]cty.Value)
+		for name, aT := range atys {
+			av := val.GetAttr(name)
+			newAV := hcty2ctyWithType(aT, av)
+			newAVs[name] = newAV
+		}
+		return cty.ObjectVal(newAVs)
+	case ty.IsCapsuleType():
+		contract.Assertf(false, "Capsule types are not yet supported")
 	}
+	contract.Assertf(false, "match failure on hcty.Value: %v", val.GoString())
+	return cty.DynamicVal
+}
 
-	typeJSON, err := ctyjson.MarshalType(v.Type())
-	if err != nil {
-		return hcty.False, err
+func ctype2htype(t cty.Type) (hcty.Type, error) {
+	if t.GoString() == "cty.NilType" {
```

Comment:
Can you leave a comment on why this is necessary? I'm not 100% sure I understand.

Reply:
Removing this panics in the turnaround test... Be my guest.. Some kind of insanity in the underlying lib I think.

Reply:
This seems less brittle and allows `go test ./... --run=TestCtyTurnaround` to pass (if thats the turnaround test your talking about).
```suggestion
	if t == hcty.NilType {
```

Reply:
t.Equals(hcty.NilType) panics.

Reply:
That makes sense, since `t` and `hcty.NilType` are from different sets of types.